File: src/Calculator/Converters/BooleanToVisibilityConverter.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "BooleanToVisibilityConverter.h"

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Interop;

namespace CalculatorApp
{
    namespace Converters
    {
        Object ^ BooleanToVisibilityConverter::Convert(Object ^ value, TypeName targetType, Object ^ parameter, String ^ language)
        {
            (void)targetType; // Unused parameter
            (void)parameter;  // Unused parameter
            (void)language;   // Unused parameter

            auto boxedBool = dynamic_cast<Box<bool> ^>(value);
            auto boolValue = (boxedBool != nullptr && boxedBool->Value);
            return BooleanToVisibilityConverter::Convert(boolValue);
        }

        Object ^ BooleanToVisibilityConverter::ConvertBack(Object ^ value, TypeName targetType, Object ^ parameter, String ^ language)
        {
            (void)targetType; // Unused parameter
            (void)parameter;  // Unused parameter
            (void)language;   // Unused parameter

            auto visibility = dynamic_cast<Box<Visibility> ^>(value);
            return (visibility != nullptr && visibility->Value == Visibility::Visible);
        }

        Object ^ BooleanToVisibilityNegationConverter::Convert(Object ^ value, TypeName targetType, Object ^ parameter, String ^ language)
        {
            (void)targetType; // Unused parameter
            (void)parameter;  // Unused parameter
            (void)language;   // Unused parameter

            auto boxedBool = dynamic_cast<Box<bool> ^>(value);
            auto boolValue = (boxedBool != nullptr && boxedBool->Value);
            return BooleanToVisibilityConverter::Convert(!boolValue);
        }

        Object ^ BooleanToVisibilityNegationConverter::ConvertBack(Object ^ value, TypeName targetType, Object ^ parameter, String ^ language)
        {
            (void)targetType; // Unused parameter
            (void)parameter;  // Unused parameter
            (void)language;   // Unused parameter

            auto visibility = dynamic_cast<Box<Visibility> ^>(value);
            return (visibility != nullptr && visibility->Value != Visibility::Visible);
        }

    }
}

File: src/Calculator/Converters/BooleanToVisibilityConverter.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;

using Windows.UI.Xaml;

namespace CalculatorApp
{
    namespace Converters
    {
        /// <summary>
        /// Value converter that translates true to <see cref="Visibility.Visible"/> and false
        /// to <see cref="Visibility.Collapsed"/>.
        /// </summary>
        public sealed class BooleanToVisibilityConverter : Windows.UI.Xaml.Data.IValueConverter
        {
            public static Windows.UI.Xaml.Visibility Convert(bool visibility)
            {
                return visibility ? Windows.UI.Xaml.Visibility.Visible : Windows.UI.Xaml.Visibility.Collapsed;
            }

            public object Convert(object value, Type targetType, object parameter, string language)
            {
                var boolValue = (value is bool boxedBool && boxedBool);
                return BooleanToVisibilityConverter.Convert(boolValue);
            }

            public object ConvertBack(object value, Type targetType, object parameter, string language)
            {
                return (value is Visibility visibility && visibility == Visibility.Visible);
            }
        }

        /// <summary>
        /// Value converter that translates false to <see cref="Visibility.Visible"/> and true
        /// to <see cref="Visibility.Collapsed"/>.
        /// </summary>
        public sealed class BooleanToVisibilityNegationConverter : Windows.UI.Xaml.Data.IValueConverter
        {
            public object Convert(object value, Type targetType, object parameter, string language)
            {
                var boolValue = (value is bool boxedBool && boxedBool);
                return BooleanToVisibilityConverter.Convert(!boolValue);
            }

            public object ConvertBack(object value, Type targetType, object parameter, string language)
            {
                return (value is Visibility visibility && visibility != Visibility.Visible);
            }
        }
    }
}

